pl_bolts/models/detection/yolo/yolo_layers.py

```diff
@@ -0,0 +1,553 @@
+from typing import Callable, Dict, List, Optional, Tuple
+
+import torch
+from pytorch_lightning.utilities.exceptions import MisconfigurationException
+from torch import nn, Tensor
+
+from pl_bolts.utils.warnings import warn_missing_pkg
+
+try:
+    from torchvision.ops import box_area, box_iou
+except ModuleNotFoundError:
+    warn_missing_pkg('torchvision')  # pragma: no-cover
+    _TORCHVISION_AVAILABLE = False
+else:
+    _TORCHVISION_AVAILABLE = True
+
+
+def _corner_coordinates(xy, wh):
+    """
+    Converts box center points and sizes to corner coordinates.
+
+    Args:
+        xy (Tensor): Center coordinates. Tensor of size `[..., 2]`.
+        wh (Tensor): Width and height. Tensor of size `[..., 2]`.
+
+    Returns:
+        boxes (Tensor): A matrix of (x1, y1, x2, y2) coordinates.
+    """
+    half_wh = wh / 2
+    top_left = xy - half_wh
+    bottom_right = xy + half_wh
+    return torch.cat((top_left, bottom_right), -1)
+
+
+def _aligned_iou(dims1, dims2):
+    """
+    Calculates a matrix of intersections over union from box dimensions, assuming that the boxes
+    are located at the same coordinates.
+
+    Args:
+        dims1 (Tensor[N, 2]): width and height of N boxes
+        dims2 (Tensor[M, 2]): width and height of M boxes
+
+    Returns:
+        iou (Tensor[N, M]): the NxM matrix containing the pairwise IoU values for every element in
+            `dims1` and `dims2`
+    """
+    area1 = dims1[:, 0] * dims1[:, 1]  # [N]
+    area2 = dims2[:, 0] * dims2[:, 1]  # [M]
+
+    inter_wh = torch.min(dims1[:, None, :], dims2)  # [N, M, 2]
+    inter = inter_wh[:, :, 0] * inter_wh[:, :, 1]  # [N, M]
+    union = area1[:, None] + area2 - inter  # [N, M]
+
+    return inter / union
+
+
+def _elementwise_iou(boxes1: Tensor, boxes2: Tensor) -> Tensor:
+    """
+    Returns the elementwise intersection-over-union between two sets of boxes.
+
+    Both sets of boxes are expected to be in (x1, y1, x2, y2) format.
+
+    Args:
+        boxes1 (Tensor[N, 4])
+        boxes2 (Tensor[N, 4])
+
+    Returns:
+        iou (Tensor[N]): the vector containing the elementwise IoU values for every element in
+        boxes1 and boxes2
+    """
+    area1 = box_area(boxes1)
+    area2 = box_area(boxes2)
+
+    lt = torch.max(boxes1[:, :2], boxes2[:, :2])  # [N,2]
+    rb = torch.min(boxes1[:, 2:], boxes2[:, 2:])  # [N,2]
+
+    wh = (rb - lt).clamp(min=0)  # [N,2]
+    inter = wh[:, 0] * wh[:, 1]  # [N]
+
+    iou = inter / (area1 + area2 - inter)
+    return iou
+
+
+def _elementwise_generalized_iou(boxes1: Tensor, boxes2: Tensor) -> Tensor:
```

Comment:
We recently added this to torchvision 0.8 release `torchvision.ops.generalized_box_iou`

Reply:
See the discussion about `_elementwise_iou` above.